In [1]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import statistics 
import math

In [2]:
matches = pd.read_csv("/content/sample_data/matches.csv")
matches.head()

id  season       city  ...         umpire1        umpire2 umpire3
0   1    2017  Hyderabad  ...     AY Dandekar       NJ Llong     NaN
1   2    2017       Pune  ...  A Nand Kishore         S Ravi     NaN
2   3    2017     Rajkot  ...     Nitin Menon      CK Nandan     NaN
3   4    2017     Indore  ...    AK Chaudhary  C Shamshuddin     NaN
4   5    2017  Bangalore  ...             NaN            NaN     NaN

[5 rows x 18 columns]

In [3]:
# setting up the table with relevant columns
dropList = ["result","player_of_match","venue","umpire1","umpire2","umpire3"]
matches.drop(labels=dropList, axis=1, inplace=True)
matches.head() 

id  season  ... win_by_runs win_by_wickets
0   1    2017  ...          35              0
1   2    2017  ...           0              7
2   3    2017  ...           0             10
3   4    2017  ...           0              6
4   5    2017  ...          15              0

[5 rows x 12 columns]

In [4]:
matches[pd.isnull(matches['winner'])]

id  season       city  ... winner win_by_runs win_by_wickets
300    301    2011      Delhi  ...    NaN           0              0
545    546    2015  Bangalore  ...    NaN           0              0
570    571    2015  Bangalore  ...    NaN           0              0
744  11340    2019  Bengaluru  ...    NaN           0              0

[4 rows x 12 columns]

In [5]:
matches['winner'].fillna('Draw', inplace=True)
matches[pd.isnull(matches['winner'])]

Empty DataFrame
Columns: [id, season, city, date, team1, team2, toss_winner, toss_decision, dl_applied, winner, win_by_runs, win_by_wickets]
Index: []

In [6]:
matches[pd.isnull(matches['city'])]

id  season city  ...                       winner win_by_runs win_by_wickets
461  462    2014  NaN  ...  Royal Challengers Bangalore           0              7
462  463    2014  NaN  ...             Delhi Daredevils           0              4
466  467    2014  NaN  ...          Chennai Super Kings           7              0
468  469    2014  NaN  ...          Sunrisers Hyderabad           4              0
469  470    2014  NaN  ...          Chennai Super Kings           0              7
474  475    2014  NaN  ...              Kings XI Punjab           0              5
476  477    2014  NaN  ...          Sunrisers Hyderabad          15              0

[7 rows x 12 columns]

In [7]:
matches['city'].fillna('Dubai', inplace=True)
matches[pd.isnull(matches['city'])]

Empty DataFrame
Columns: [id, season, city, date, team1, team2, toss_winner, toss_decision, dl_applied, winner, win_by_runs, win_by_wickets]
Index: []

In [8]:
matches.replace(["Deccan Chargers","Delhi Daredevils"],["Sunrisers Hyderabad","Delhi Capitals"],inplace=True,regex=True)
match1 = matches[((matches.team1=="Kolkata Knight Riders") & (matches.team2=="Rajasthan Royals")) | ((matches.team1=="Rajasthan Royals") & (matches.team2=="Kolkata Knight Riders"))]
match1.shape[0]

20

In [9]:
mw_kkr = 0 
mw_rr = 0
lst= [i for i in match1['winner']] 
print("Win Tracker!")
for i in lst:
  if i=="Rajasthan Royals":
    mw_rr += 1
  elif i=='Draw':
    continue
  else:
    mw_kkr += 1
  print(str(mw_kkr)+" "+str(mw_rr))
print("KKR vs RR : "+str(mw_kkr)+" "+str(mw_rr))

Win Tracker!
0 1
0 2
0 3
1 3
1 4
2 4
3 4
4 4
4 5
5 5
5 6
6 6
6 7
6 8
6 9
7 9
8 9
9 9
10 9
10 10
KKR vs RR : 10 10


In [10]:
last_3_season = match1[match1.season >= 2017]
last_3_season.groupby('winner').winner.count()

winner
Kolkata Knight Riders    4
Rajasthan Royals         1
Name: winner, dtype: int64

Out of 20 matches held between KKR and RR compared in two scenarios. Past three seasons record states that KKR is in clear lead 4 to 1 against RR.But IPL record has been equal (i.e.) 10 is to 10 

In [11]:
def statistics_for_lists(lst):
  print("Maximum Value Of List:")
  print(max(lst))
  print("Median of the List:")
  print(statistics.median(lst))
  print("Mean of the List:")
  print(statistics.mean(lst))
  print("75% of the Median is:")
  print(statistics.median_high(lst))
  print("Minimum Value of List:")
  print(min(lst))

In [12]:
deliveries = pd.read_csv("/content/sample_data/deliveries.csv")
deliveries.head()

match_id  inning  ... dismissal_kind fielder
0         1       1  ...            NaN     NaN
1         1       1  ...            NaN     NaN
2         1       1  ...            NaN     NaN
3         1       1  ...            NaN     NaN
4         1       1  ...            NaN     NaN

[5 rows x 21 columns]

In [13]:
dropToBeList = ['inning','is_super_over','bye_runs','legbye_runs','fielder']
deliveries.drop(dropToBeList, axis=1, inplace=True)
deliveries.replace(['Deccan Chargers','Delhi Daredevils'],['Sunrisers Hyderabad','Delhi Capitals'],inplace=True,regex=True)
deliveries['dismissal_kind'].fillna('Not Out',inplace=True)
deliveries.head()

match_id         batting_team  ... player_dismissed  dismissal_kind
0         1  Sunrisers Hyderabad  ...              NaN         Not Out
1         1  Sunrisers Hyderabad  ...              NaN         Not Out
2         1  Sunrisers Hyderabad  ...              NaN         Not Out
3         1  Sunrisers Hyderabad  ...              NaN         Not Out
4         1  Sunrisers Hyderabad  ...              NaN         Not Out

[5 rows x 16 columns]

In [14]:
ballbyball = deliveries[((deliveries.batting_team=="Rajasthan Royals") & (deliveries.bowling_team=="Kolkata Knight Riders")) | ((deliveries.batting_team=="Kolkata Knight Riders") & (deliveries.bowling_team=="Rajasthan Royals"))]
no_of_matches=list(set([i for i in ballbyball['match_id']]))
no_of_matches.sort()
print(len(no_of_matches))

20


In [15]:
#Q2
nof_wickets = ballbyball.groupby('match_id').player_dismissed.count()
nof_wickets.describe()

count    20.000000
mean     12.200000
std       3.334035
min       5.000000
25%      11.000000
50%      12.000000
75%      15.000000
max      17.000000
Name: player_dismissed, dtype: float64

There is a possibility of wickets falling between 12 to 15


In [16]:
#Q4
jofra = deliveries[deliveries.bowler=='J Archer']
runs_given = jofra.groupby('match_id').total_runs.sum().reset_index()
lst_runs_given = [i for i in runs_given['total_runs']]
nof_over_per_match = jofra.groupby('match_id').over.nunique().reset_index()
over_per_match_id = [i for i in nof_over_per_match['over']]
economy = []
for i in range(len(lst_runs_given)):
  val = lst_runs_given[i]/over_per_match_id[i]
  economy.append(val)
statistics_for_lists(economy)
no_of_4s = jofra[jofra.batsman_runs==4].groupby('match_id').batsman_runs.count()
no_of_4s.describe()

Maximum Value Of List:
13.0
Median of the List:
7.75
Mean of the List:
8.11111111111111
75% of the Median is:
7.75
Minimum Value of List:
4.25


count    18.000000
mean      3.055556
std       1.551933
min       1.000000
25%       2.000000
50%       3.000000
75%       4.000000
max       6.000000
Name: batsman_runs, dtype: float64

In [17]:
#Q4
jofra = ballbyball[ballbyball.bowler=='J Archer']
runs_given = jofra.groupby('match_id').total_runs.sum().reset_index()
lst_runs_given = [i for i in runs_given['total_runs']]
nof_over_per_match = jofra.groupby('match_id').over.nunique().reset_index()
over_per_match_id = [i for i in nof_over_per_match['over']]
economy = []
for i in range(len(lst_runs_given)):
  val = lst_runs_given[i]/over_per_match_id[i]
  economy.append(val)
statistics_for_lists(economy)
no_of_4s = jofra[jofra.batsman_runs==4].groupby('match_id').batsman_runs.count()
no_of_4s.describe()

Maximum Value Of List:
10.75
Median of the List:
8.25
Mean of the List:
8.145833333333334
75% of the Median is:
9.25
Minimum Value of List:
5.333333333333333


count    3.000000
mean     3.000000
std      2.645751
min      1.000000
25%      1.500000
50%      2.000000
75%      4.000000
max      6.000000
Name: batsman_runs, dtype: float64

Jofra has an economy rate of 8.25 over KKR and 7.75 across IPL. He is somewhat expensive over KKR than that of over teams in IPL. He may give away 2 or 3 fours on an average against KKR and same is the case in IPL

In [19]:
#Q5
economy_lessthan_8=[]
for i in no_of_matches:
  df = ballbyball[ballbyball.match_id==i]
  count=0
  lst_bowler = list(set([i for i in df['bowler']]))
  for j in lst_bowler:
    stats = df[df.bowler==j]
    runs_given = stats['total_runs'].sum()
    overs_bowled = stats['over'].nunique()
    if(overs_bowled>1):
      economy = runs_given/overs_bowled
      if(economy<8):
        count += 1
  economy_lessthan_8.append(count)
statistics_for_lists(economy_lessthan_8)

Maximum Value Of List:
11
Median of the List:
6.0
Mean of the List:
6.2
75% of the Median is:
6
Minimum Value of List:
1


There are 6 bowlers present in the match between KKR and RR who has an economy less than 8

In [41]:
lst_batsman = list(set([i for i in ballbyball['batsman']]))
ratios=[]
for i in lst_batsman:
  lst=[]
  df = ballbyball[ballbyball.batsman==i]
  balls_faced = df[(df.wide_runs==0) | (df.noball_runs==0)].groupby('match_id').ball.count().reset_index()
  total_runs_per_match = df.groupby('match_id').batsman_runs.sum().reset_index()
  total_runs_per_match.rename(columns={'batsman_runs':'total_runs'},inplace=True)
  score_4 = df[df.batsman_runs==4].groupby('match_id').batsman_runs.sum().reset_index()
  score_4.rename(columns={'batsman_runs':'score_4'},inplace=True)
  score_123 = df[(df.batsman_runs==1) | (df.batsman_runs==2) | (df.batsman_runs==3)].groupby('match_id').batsman_runs.sum().reset_index()
  score_123.rename(columns={'batsman_runs':'score_123'},inplace=True)
  d1 = pd.merge(total_runs_per_match, score_4, how='outer')
  d2 = pd.merge(score_123, balls_faced, how='outer')
  d = pd.merge(d1, d2, how='outer')
  d.fillna(0,inplace=True)
  d = d[d.ball>10]
  if(d.shape[0]==0):
    ratios.append(0)
  else:
    lst_tot_runs = [i for i in d['total_runs']]
    lst_score_4 = [i for i in d['score_4']]
    lst_score_123 = [i for i in d['score_123']]
    for i in range(len(lst_tot_runs)):
      val = (lst_score_4[i] - lst_score_123[i])/lst_tot_runs[i]
      lst.append(val)
    ratios.append(statistics.median(lst))
for i,j in zip(lst_batsman, ratios):
  print(i+" "+str(round(j,2)))

Mohammad Hafeez 0
Niraj Patel -0.43
KK Cooper 0
Anureet Singh 0
SW Tait 0
SR Watson 0.16
I Sharma 0
PC Valthaty 0
S Sreesanth 0
AB Agarkar -0.5
BB McCullum 0
MK Tiwary -0.05
S Gill 0.07
LRPL Taylor -0.2
DB Das -0.35
RA Jadeja -0.43
JH Kallis -0.28
Umar Gul 0
FY Fazal 0.08
AC Voges -0.24
P Dogra -1.0
AJ Finch -0.71
SK Trivedi 0
AL Menaria -0.19
CH Morris 0
MS Bisla -0.17
SPD Smith -0.22
K Gowtham 0
RJ Quiney 0
CA Pujara 0.06
SP Goswami -0.36
N Rana 0.0
J Archer 0.0
AG Paunikar 0.42
KK Nair -0.14
B Lee 0.12
PP Chawla 0
BJ Hodge -0.26
DH Yagnik -0.12
RA Tripathi 0.07
UT Yadav 0.58
WP Saha 0
Salman Butt 0.67
JP Faulkner 0
SB Bangar 0
A Singh 0
AA Jhunjhunwala -0.11
CR Brathwaite 0
NV Ojha 0.33
M Morkel 0
RN ten Doeschate 0
SP Narine 0.27
LR Shukla -0.32
M Rawat 0
EJG Morgan -0.18
AD Russell 0.04
MK Pandey -0.06
JC Buttler 0.18
T Henderson 0
J Botha -1.0
R Singh 0
R Bhatia 0
J Searles 0
AM Rahane -0.03
GC Smith -0.07
SV Samson -0.35
R Dravid 0.01
D Short 0.02
BA Stokes -1.0
I Sodhi 0
CH Gay